clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java

```diff
@@ -163,127 +159,179 @@ private boolean allSubscriptionsEqual(Set<String> allTopics,
             List<TopicPartition> ownedPartitions = consumerEntry.getValue();
 
             List<TopicPartition> consumerAssignment = assignment.get(consumer);
-            int i = 0;
-            // assign the first N partitions up to the max quota, and mark the remaining as being revoked
-            for (TopicPartition tp : ownedPartitions) {
-                if (i < maxQuota) {
-                    consumerAssignment.add(tp);
-                    unassignedPartitions.remove(tp);
-                } else {
-                    allRevokedPartitions.add(tp);
-                }
-                ++i;
-            }
 
             if (ownedPartitions.size() < minQuota) {
+                // the expected assignment size is more than consumer have now, so keep all the owned partitions
+                // and put this member into unfilled member list
+                if (ownedPartitions.size() > 0) {
+                    consumerAssignment.addAll(ownedPartitions);
+                    toBeRemovedPartitions.addAll(ownedPartitions);
+                }
                 unfilledMembers.add(consumer);
+            } else if (ownedPartitions.size() >= maxQuota && numMaxCapacityMembers++ <= numExpectedMaxCapacityMembers) {
+                // consumer owned the "maxQuota" of partitions or more, and we still under the number of expected max capacity members
+                // so keep "maxQuota" of the owned partitions, and revoke the rest of the partitions
+                consumerAssignment.addAll(ownedPartitions.subList(0, maxQuota));
+                toBeRemovedPartitions.addAll(ownedPartitions.subList(0, maxQuota));
+                allRevokedPartitions.addAll(ownedPartitions.subList(maxQuota, ownedPartitions.size()));
             } else {
-                // It's possible for a consumer to be at both min and max capacity if minQuota == maxQuota
-                if (consumerAssignment.size() == minQuota)
-                    minCapacityMembers.add(consumer);
-                if (consumerAssignment.size() == maxQuota)
-                    maxCapacityMembers.add(consumer);
+                // consumer owned the "minQuota" of partitions or more
+                // so keep "minQuota" of the owned partitions, and revoke the rest of the partitions
+                consumerAssignment.addAll(ownedPartitions.subList(0, minQuota));
+                toBeRemovedPartitions.addAll(ownedPartitions.subList(0, minQuota));
+                allRevokedPartitions.addAll(ownedPartitions.subList(minQuota, ownedPartitions.size()));
             }
         }
 
+        List<TopicPartition> unassignedPartitions;
+        if (!toBeRemovedPartitions.isEmpty()) {
+            Collections.sort(toBeRemovedPartitions,
+                Comparator.comparing(TopicPartition::topic).thenComparing(TopicPartition::partition));
+            unassignedPartitions = getUnassignedPartitions(sortedAllPartitions, toBeRemovedPartitions);
+            sortedAllPartitions = null;
```

Comment:
`sortedAllPartitions` can be cleared for both cases I think, like `toBeRemovedPartitions`?

Reply:
I've deleted `sortedAllPartitions`, so no need to delete it now.